### PR TITLE
JDK17 adopts RI java.lang.String

### PR DESCRIPTION
**Pull request by JasonFengJ9**
Not include OpenJ9 `java.lang.String` for `JDK17`;
Update `JDK17` `StringCoding` references which have been moved into `String`;
Support `ISO-8859-1`/`Latin-1` characters between `0x80` and `0xFF` in String compression mode which is required by RI `String`.

~depends https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/295~ - **merged**
unlock https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/296 https://github.com/eclipse/openj9/pull/12547

fixes https://github.com/eclipse/openj9/issues/12398

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

### CODE REVIEW

**Comment by JasonFengJ9** [comment]
Jenkins copyright check

File: runtime/oti/VMHelpers.hpp [hunk 1]
```diff
@@ -1015,6 +1015,40 @@ class VM_VMHelpers
...
+	/**
+	 * Check if the UTF8 byte stream contains only ISO-8859-1/Latin-1 characters.
+	 *
+	 * @param data[in] the UTF8 byte stream
+	 * @param length[in] the length of incoming UTF8 byte stream
+	 *
+	 * @returns true if data only contains Latin-1 characters, false if otherwise
+	 */
+	static VMINLINE bool
+	isLatin1String(const U_8 *data, const UDATA length)
```
Comment (fjeremic) [inline runtime/oti/VMHelpers.hpp:1028]: Are we able to sue this in `j9gc_createJavaLangString` to eliminate duplicated code? I think there are other places we check for whether strings are compressible. I thought we had an API for this already as well. Might be worth checking so as to avoid duplication of code.

Comment (JasonFengJ9) [reply runtime/oti/VMHelpers.hpp:1028]: This helper method is used by `j9gc_createJavaLangString`, my thought was that `VHelpers.hpp` is a better home if there is a reference from other places.

> I think there are other places we check for whether strings are compressible.

Are they `IS_STRING_COMPRESSED` and `IS_STRING_COMPRESSED_VM` which take a string object? This method expects a `U_8* data` instead.

Comment (JasonFengJ9) [reply runtime/oti/VMHelpers.hpp:1028]: @fjeremic ping

Comment (fjeremic) [reply runtime/oti/VMHelpers.hpp:1028]: > Are they IS_STRING_COMPRESSED and IS_STRING_COMPRESSED_VM which take a string object? This method expects a U_8* data instead.

These macros only check the `coder` field. Just checked `j9gc_createJavaLangString` again. Looks ok to me.

**Review by fjeremic — APPROVED** [review]

**Comment by JasonFengJ9** [comment]
@pshipton could you review & merge if approved?

**Comment by pshipton** [comment]
@gacholio @dmitripivkine pls review the VM/GC c code.

File: runtime/gc_base/StringTable.cpp [hunk 2]
```diff
@@ -568,6 +604,10 @@ j9gc_createJavaLangString(J9VMThread *vmThread, U_8 *data, UDATA length, UDATA s
 			}
 		}
 	}
+	bool isLatin1 = isASCII;
+	if (compressStrings && J2SE_VERSION(vm) >= J2SE_V17) {
```
Comment (pshipton) [inline runtime/gc_base/StringTable.cpp:608]: This doesn't seem the right place for this check, it doesn't take isUnicode into account. Also I think the check for isLatin1String can be incorporated into the check for isASCII, to avoid scanning the input twice.

File: runtime/gc_base/StringTable.cpp [hunk 3]
```diff
@@ -568,6 +604,10 @@ j9gc_createJavaLangString(J9VMThread *vmThread, U_8 *data, UDATA length, UDATA s
 			}
 		}
 	}
+	bool isLatin1 = isASCII;
```
Comment (pshipton) [inline runtime/gc_base/StringTable.cpp:607]: Since this variable is used differently for < 17 and >= 17, I think it needs a better name, such as isASCIIorLatin1.

File: runtime/gc_base/StringTable.cpp (hunk 2 above)
Comment (pshipton) [reply runtime/gc_base/StringTable.cpp:608]: i.e. we only need to decode UTF8 which isn't already determined to be ASCII.

Comment (JasonFengJ9) [reply runtime/gc_base/StringTable.cpp]: Moved the code in question, check `isASCII` first, then check `isASCIIorLatin1` if required to minimize the overhead for pre-JDK17 levels, also added the check for `isUnicode` case.

File: runtime/gc_base/StringTable.cpp (hunk 3 above)
Comment (JasonFengJ9) [reply runtime/gc_base/StringTable.cpp]: Updated.

File: runtime/gc_base/StringTable.cpp [hunk 4]
```diff
@@ -666,20 +706,29 @@ j9gc_createJavaLangString(J9VMThread *vmThread, U_8 *data, UDATA length, UDATA s
...
 				if (translateSlashes) {
-					for (UDATA i = 0; i < length; ++i) {
-						U_8 c = data[i];
-						if ('/' == c) {
-							lastSlash = i;
-							c = '.';
+					if (isASCII) {
+						for (UDATA i = 0; i < length; ++i) {
+							U_8 c = data[i];
+							c = data[i];
```
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply runtime/gc_base/StringTable.cpp]: Removed.

File: runtime/compiler/env/j9method.cpp [hunk 5]
```diff
@@ -3637,6 +3649,7 @@ void TR_ResolvedJ9Method::construct()
       {x(TR::java_lang_Character_isLowerCase,         "isLowerCase",          "(I)Z")},
       {x(TR::java_lang_Character_isWhitespace,        "isWhitespace",         "(I)Z")},
       {x(TR::java_lang_Character_isAlphabetic,        "isAlphabetic",         "(I)Z")},
+	  {x(TR::java_lang_Character_toLowerCase,         "toLowerCase",          "(I)I")},
```
Comment (keithc-ca) [inline runtime/compiler/env/j9method.cpp:3652]: Inconsistent indentation.

Comment (JasonFengJ9) [reply runtime/compiler/env/j9method.cpp]: Fixed.

File: runtime/gc_base/StringTable.cpp [hunk 6]
```diff
@@ -556,9 +594,19 @@ j9gc_createJavaLangString(J9VMThread *vmThread, U_8 *data, UDATA length, UDATA s
 					break;
 				}
 			}
+			isASCIIorLatin1 = isASCII;
+			if (!isASCII && compressStrings && J2SE_VERSION(vm) >= J2SE_V17) {
```
Comment (keithc-ca) [inline runtime/gc_base/StringTable.cpp:598]: The test of `compressStrings` is redundant here (see line 589).

File: runtime/gc_base/StringTable.cpp [hunk 7]
```diff
@@ -666,20 +718,28 @@ j9gc_createJavaLangString(J9VMThread *vmThread, U_8 *data, UDATA length, UDATA s
...
-					for (UDATA i = 0; i < length; ++i) {
-						J9JAVAARRAYOFBYTE_STORE(vmThread, charArray, i, data[i]);
+					if (isASCII) {
+						for (UDATA i = 0; i < length; ++i) {
+							J9JAVAARRAYOFBYTE_STORE(vmThread, charArray, i, data[i]);
+						}
+					} else {
+						lastSlash = storeLatin1ByteArrayhelper(vmThread, data, length, charArray, false);
 					}
 				}
```
Comment (keithc-ca) [inline runtime/gc_base/StringTable.cpp:744]: Please help me understand why `charArray` actually refers to a byte array in this block; it doesn't appear to be guarded by `J9_ARE_ANY_BITS_SET(vm->runtimeFlags, J9_RUNTIME_STRING_BYTE_ARRAY)`.

Comment (gacholio) [reply runtime/gc_base/StringTable.cpp:744]: This is a bit of a hack - the primitive array macros work on any primitive array, so if we're storing bytes (either in a byte or char array), the byte macros are used.

Comment (keithc-ca) [reply runtime/gc_base/StringTable.cpp:744]: I agree with the 'hack' label, but thanks for the explanation.

File: runtime/gc_base/StringTable.cpp (hunk 6 above)
Comment (JasonFengJ9) [reply runtime/gc_base/StringTable.cpp]: Fixed.

File: runtime/gc_base/StringTable.cpp [hunk 8]
```diff
@@ -556,9 +594,19 @@ j9gc_createJavaLangString(J9VMThread *vmThread, U_8 *data, UDATA length, UDATA s
 					break;
 				}
 			}
+			isASCIIorLatin1 = isASCII;
+			if (!isASCII && J2SE_VERSION(vm) >= J2SE_V17) {
+				for (UDATA i = 0; i < length; ++i) {
```
Comment (pshipton) [inline runtime/gc_base/StringTable.cpp:599]: This shouldn't loop over the data twice. Above we check for characters > 0x7F, this can be extended to also check for > 0xFF.

File: runtime/gc_base/StringTable.cpp [hunk 9]
```diff
@@ -524,6 +524,42 @@ j9gc_stringHashFn(void *key, void *userData)
...
+	UDATA lengthTmp = length;
+	UDATA index = 0;
+	while (0 != lengthTmp) {
+		U_16 unicode = 0;
+		UDATA consumed = VM_VMHelpers::decodeUTF8CharN(dataTmp, &unicode, lengthTmp);
+		if (0 == consumed) {
+			Assert_MM_unreachable();
+		}
+		dataTmp += consumed;
+		lengthTmp -= consumed;
```
Comment (dmitripivkine) [inline runtime/gc_base/StringTable.cpp:552]: If it possible (not too much performance impact) I would add check `(consumed <= lengthTmp)` before subtraction and assert otherwise. `lengthTmp` is declared unsigned so it might be wrapped around and negative result treated as very large positive. This might be a reaction (for example) to garbage passed and make execution go out of control

File: runtime/gc_base/StringTable.cpp [hunk 10]
```diff
@@ -567,6 +615,10 @@ j9gc_createJavaLangString(J9VMThread *vmThread, U_8 *data, UDATA length, UDATA s
 				break;
 			}
 		}
+		isASCIIorLatin1 = isASCII;
+		if (!isASCII && compressStrings && J2SE_VERSION(vm) >= J2SE_V17) {
```
Comment (pshipton) [inline runtime/gc_base/StringTable.cpp:619]: Like above, this shouldn't loop over the data twice. We can extend the above check for > 0x7F to check the UTF8 characters.

File: runtime/gc_base/StringTable.cpp (hunk 9 above)
Comment (gacholio) [reply runtime/gc_base/StringTable.cpp:552]: The assumption here is that the input data is valid, so even the assert is somewhat unnecessary. Also note that the tmp variables are unnecessary (the args can be modified in place).

File: runtime/gc_base/StringTable.cpp [hunk 11]
```diff
@@ -556,9 +594,19 @@ j9gc_createJavaLangString(J9VMThread *vmThread, U_8 *data, UDATA length, UDATA s
 					break;
 				}
 			}
+			isASCIIorLatin1 = isASCII;
+			if (!isASCII && J2SE_VERSION(vm) >= J2SE_V17) {
```
Comment (gacholio) [inline runtime/gc_base/StringTable.cpp:598]: Please bracket the second test (here and everywhere else).

File: runtime/gc_base/StringTable.cpp (hunk 8 above)
Comment (pshipton) [reply runtime/gc_base/StringTable.cpp:599]: Another way to look at it, this loop can start at the first non-ascii character as found in the check above, it doesn't need to start again from zero. Similarly below.

File: runtime/gc_base/StringTable.cpp (hunk 9 above)
Comment (JasonFengJ9) [reply runtime/gc_base/StringTable.cpp:552]: Removing the tmp variables and use the args instead.

File: runtime/gc_base/StringTable.cpp (hunk 8 above)
Comment (JasonFengJ9) [reply runtime/gc_base/StringTable.cpp:599]: I didn't do this way to avoid introducing extra overhead for pre-JDK17 levels. I am fine with either way.

File: runtime/gc_base/StringTable.cpp (hunk 10 above)
Comment (JasonFengJ9) [reply runtime/gc_base/StringTable.cpp:619]: This is a slightly different case, for UTF8 byte stream it is not straightforward to check if it is greater than `0xFF`, `isLatin1String:decodeUTF8CharN()` is required.

File: runtime/gc_base/StringTable.cpp (hunk 8 above)
Comment (JasonFengJ9) [reply runtime/gc_base/StringTable.cpp:599]: > Another way to look at it, this loop can start at the first non-ascii character as found in the check above, it doesn't need to start again from zero. Similarly below.

Agreed.

Comment (keithc-ca) [reply runtime/gc_base/StringTable.cpp:599]: Or compute the max char value, exiting the loop if it exceeds 0xFF.
If that max is not more than 0x7F, it is ASCII; if no more than 0xFF it is Latin1.

Comment (JasonFengJ9) [reply runtime/gc_base/StringTable.cpp:599]: Technically that introduces an overhead for pre-JDK17 levels which initially might exit the loop as early as first byte check, or even for JDK17+ as well.

Comment (keithc-ca) [reply runtime/gc_base/StringTable.cpp]: The upper bound (0xFF or 0x7F) can be set based on JDK version.

File: runtime/gc_base/StringTable.cpp (hunk 11 above)
Comment (JasonFengJ9) [reply runtime/gc_base/StringTable.cpp]: Added the bracket.

File: runtime/gc_base/StringTable.cpp (hunk 9 above)
Comment (JasonFengJ9) [reply runtime/gc_base/StringTable.cpp]: Updated the assert like https://github.com/eclipse/openj9/blob/8317935a0b1d067d4a67b8c0378a20dfcbf4d87e/runtime/gc_base/StringTable.cpp#L626

File: runtime/gc_base/StringTable.cpp (hunk 8 above)
Comment (JasonFengJ9) [reply runtime/gc_base/StringTable.cpp]: Moved `isASCIIorLatin1` computation into `isASCII` loop.

File: runtime/gc_base/StringTable.cpp (hunk 10 above)
Comment (JasonFengJ9) [reply runtime/gc_base/StringTable.cpp]: Moved the code in question similarly.

File: runtime/gc_base/StringTable.cpp [hunk 12]
```diff
@@ -553,17 +587,31 @@ j9gc_createJavaLangString(J9VMThread *vmThread, U_8 *data, UDATA length, UDATA s
 			for (UDATA i = 0; i < length; ++i) {
 				if (unicodeData[i] > 0x7F) {
 					isASCII = false;
+					if (J2SE_VERSION(vm) >= J2SE_V17) {
+						for (UDATA j = i; j < length; ++j) {
+							if ((unicodeData[j] > 0xFF)) {
```
Comment (keithc-ca) [inline runtime/gc_base/StringTable.cpp:592]: Too many parentheses here.

Comment (JasonFengJ9) [reply runtime/gc_base/StringTable.cpp]: Removed extra one.

**Comment by pshipton** [comment]
jenkins test sanity zlinux jdk8,jdknext

**Comment by pshipton** [comment]
jenkins compile win jdk8

**Comment by pshipton** [comment]
There are jdk8 failures.

**Comment by JasonFengJ9** [comment]
> There are jdk8 failures.

Investigating..

**Comment by JasonFengJ9** [comment]
Missed one corner case to match `isASCIIorLatin1` value to `isASCII`.
Fixed.

**Comment by JasonFengJ9** [comment]
jenkins test sanity zlinux jdk8

**Comment by fjeremic** [comment]
Jenkins test sanity zlinux jdk8

File: runtime/gc_base/StringTable.cpp [hunk 13]
```diff
@@ -553,17 +587,33 @@ j9gc_createJavaLangString(J9VMThread *vmThread, U_8 *data, UDATA length, UDATA s
...
 	} else {
 		for (UDATA i = 0; i < length; ++i) {
 			if (data[i] > 0x7F) {
 				isASCII = false;
+				if (compressStrings && (J2SE_VERSION(vm) >= J2SE_V17)) {
+					U_8 *dataTmp = data + i;
+					UDATA lengthTmp = length - i;
+					isASCIIorLatin1 = VM_VMHelpers::isLatin1String(dataTmp, lengthTmp);
+				} else {
+					isASCIIorLatin1 = false;
```
Comment (pshipton) [inline runtime/gc_base/StringTable.cpp:615]: The unicode case needs the same fix.

Comment (JasonFengJ9) [reply runtime/gc_base/StringTable.cpp:617]: Fixed.

**Comment by pshipton** [comment]
@gacholio @keithc-ca @dmitripivkine any further concerns? 
The PR testing passed, I think this is ready to merge.

**Review by pshipton — APPROVED** [review]

**Review by dmitripivkine — APPROVED** [review]

**Review by keithc-ca — APPROVED** [review]

File: runtime/gc_base/StringTable.cpp [hunk 14]
```diff
@@ -524,6 +524,38 @@ j9gc_stringHashFn(void *key, void *userData)
...
+ * @param vmThread[in] the VM thread
+ * @param data[in] the UTF8 byte stream
+ * @param length[in] the length of incoming UTF8 byte stream
+ * @param byteArray[out] the target byte array to store the data
+ * @param translateSlashes[in] the flag to translate slashes
+ * 
+ * @return last slash position to be restored to `/` for anonymous classes
+ */
+MMINLINE static UDATA
+storeLatin1ByteArrayhelper(J9VMThread *vmThread, U_8 *data, UDATA length, j9object_t byteArray, bool translateSlashes) {
```
Comment (gacholio) [inline runtime/gc_base/StringTable.cpp:540]: { on new line.

File: runtime/oti/VMHelpers.hpp [hunk 15]
```diff
@@ -1015,6 +1015,40 @@ class VM_VMHelpers
 		return consumed;
 	}
 
+
```
Comment (gacholio) [inline runtime/oti/VMHelpers.hpp:1018]: Extra blank line

File: runtime/oti/VMHelpers.hpp [hunk 16]
```diff
@@ -1015,6 +1015,40 @@ class VM_VMHelpers
...
+	 * @param data[in] the UTF8 byte stream
+	 * @param length[in] the length of incoming UTF8 byte stream
+	 *
+	 * @returns true if data only contains Latin-1 characters, false if otherwise
+	 */
+	static VMINLINE bool
+	isLatin1String(const U_8 *data, const UDATA length)
+	{
+		bool isLatin1 = true;
+		UDATA tempLength = length;
```
Comment (gacholio) [inline runtime/oti/VMHelpers.hpp:1031]: Temps are unnecessary.

File: runtime/gc_base/StringTable.cpp (hunk 14 above)
Comment (JasonFengJ9) [reply runtime/gc_base/StringTable.cpp]: Fixed.

File: runtime/oti/VMHelpers.hpp (hunk 15 above)
Comment (JasonFengJ9) [reply runtime/oti/VMHelpers.hpp]: Removed.

File: runtime/oti/VMHelpers.hpp (hunk 16 above)
Comment (JasonFengJ9) [reply runtime/oti/VMHelpers.hpp]: Replaced the temps.

**Comment by gacholio** [comment]
jenkins compile win jdk8,jdknext

**Comment by JasonFengJ9** [comment]
The copyright check failure is an infra issue.
```
13:05:39  ERROR: Error fetching remote repo 'origin'
13:05:39  hudson.plugins.git.GitException: Failed to fetch from https://github.com/eclipse/openj9.git
13:05:39  	at hudson.plugins.git.GitSCM.fetchFrom(GitSCM.java:999)
13:05:39  	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1240)
13:05:39  	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1300)
13:05:39  	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:125)
13:05:39  	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:93)
13:05:39  	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:80)
13:05:39  	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
13:05:39  	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution$$Lambda$616/0x0000000040060080.run(Unknown Source)
13:05:39  	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
13:05:39  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
13:05:39  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
13:05:39  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
13:05:39  	at java.base/java.lang.Thread.run(Thread.java:836)
13:05:39  Caused by: hudson.plugins.git.GitException: Command "git fetch --tags --progress -- https://github.com/eclipse/openj9.git +refs/pull/12209/*:refs/remotes/origin/pr/12209/* +refs/heads/master:refs/remotes/origin/master" returned status code 128:
```

**Review by gacholio — APPROVED** [review]

**Comment by pshipton** [comment]
jenkins copyright check

**Comment by gacholio** [comment]
jenkins copyright check
